test/jdk/java/net/httpclient/CancelStreamedBodyTest.java

```diff
@@ -0,0 +1,432 @@
+
+        http2TestServer = HttpTestServer.of(new Http2TestServer("localhost", false, 0));
+        http2TestServer.addHandler(h2_chunkedHandler, "/http2/x/");
+        http2URI = "http://" + http2TestServer.serverAuthority() + "/http2/x/";
+
+        https2TestServer = HttpTestServer.of(new Http2TestServer("localhost", true, sslContext));
+        https2TestServer.addHandler(h2_chunkedHandler, "/https2/x/");
+        https2URI = "https://" + https2TestServer.serverAuthority() + "/https2/x/";
+
+        serverCount.addAndGet(4);
+        httpTestServer.start();
+        httpsTestServer.start();
+        http2TestServer.start();
+        https2TestServer.start();
+    }
+
+    @AfterTest
+    public void teardown() throws Exception {
+        String sharedClientName =
+                sharedClient == null ? null : sharedClient.toString();
+        TRACKER.track(sharedClient);
+        sharedClient = null;
+        Thread.sleep(100);
+        AssertionError fail = TRACKER.check(500);
+        try {
+            httpTestServer.stop();
+            httpsTestServer.stop();
+            http2TestServer.stop();
+            https2TestServer.stop();
+        } finally {
+            if (fail != null) {
+                if (sharedClientName != null) {
+                    System.err.println("Shared client name is: " + sharedClientName);
+                }
+                throw fail;
+            }
+        }
+    }
+
+    /**
+     * A handler that slowly sends back a body to give time for the
+     * the request to get cancelled before the body is fully received.
+     */
+    static class HTTPSlowHandler implements HttpTestHandler {
+        @Override
+        public void handle(HttpTestExchange t) throws IOException {
+            try {
+                out.println("HTTPSlowHandler received request to " + t.getRequestURI());
+                System.err.println("HTTPSlowHandler received request to " + t.getRequestURI());
+
+                byte[] req;
+                try (InputStream is = t.getRequestBody()) {
+                    req = is.readAllBytes();
+                }
+                t.sendResponseHeaders(200, -1); // chunked/variable
+                try (OutputStream os = t.getResponseBody()) {
+                    // lets split the response in several chunks...
+                    String msg = (req != null && req.length != 0)
+                            ? new String(req, UTF_8)
+                            : BODY;
+                    String[] str = msg.split("\\|");
+                    for (var s : str) {
+                        req = s.getBytes(UTF_8);
+                        os.write(req);
+                        os.flush();
+                        try {
+                            Thread.sleep(SERVER_LATENCY);
+                        } catch (InterruptedException x) {
+                            // OK
+                        }
+                        out.printf("Server wrote %d bytes%n", req.length);
```

Comment:
I think it would be better to move this printf before the call to sleep, since that's when the write completes.